File: lib/Controller/NotificationController.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:banca_movil_final/Model/Notification.dart';

class NotificationController {
  static const String _apiUrl = "localhost:9092";

  // Obtener notificaciones de un usuario
  static Future<List<NotificationI>?> getUserNotifications(int userId) async {
    var url = Uri.http(_apiUrl, "notifications/$userId");
    try {
      var response = await http.get(url);
      if (response.statusCode == 200) {
        final jsonResponse = jsonDecode(utf8.decode(response.bodyBytes)) as List;
        return jsonResponse.map((e) => NotificationI.fromJson(e)).toList();
      }
    } catch (e) {
      print(e);
    }
    return null;
  }

  // Marcar una notificación como leída
  static Future<bool> markNotificationAsRead(int notificationId) async {
    var url = Uri.http(_apiUrl, "notifications/mark-as-read/$notificationId");
    try {
      var response = await http.put(url);
      return response.statusCode == 204;
    } catch (e) {
      print(e);
    }
    return false;
  }

  // Contar notificaciones no leídas de un usuario
  static Future<int?> countNotificationNoRead(int userId) async {
    var url = Uri.http(_apiUrl, "notifications/countNotificationNoRead/$userId");
    try {
      var response = await http.get(url);
      if (response.statusCode == 200) {
        return int.parse(response.body);
      }
    } catch (e) {
      print(e);
    }
    return null;
  }

  // Marcar todas las notificaciones como leídas
  static Future<bool> markAllNotificationsAsRead(int userId) async {
    var url = Uri.http(_apiUrl, "notifications/mark-all-as-read/$userId");
    try {
      var response = await http.put(url);
      return response.statusCode == 204;
    } catch (e) {
      print(e);
    }
    return false;
  }
}

File: demo/src/main/java/com/Banca/Movil/demo/controller/NotificationController.java
package com.Banca.Movil.demo.controller;

import com.Banca.Movil.demo.model.Notification;
import com.Banca.Movil.demo.service.NotificationService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/notifications")
@CrossOrigin(origins = "*")
public class NotificationController {

    @Autowired
    private NotificationService notificationService;

    @GetMapping("/{userId}")
    public ResponseEntity<List<Notification>> getUserNotifications(@PathVariable Long userId) {
        return ResponseEntity.ok(notificationService.getUserNotifications(userId));
    }

    @PutMapping("/mark-as-read/{notificationId}")
    public ResponseEntity<Void> markNotificationAsRead(@PathVariable Long notificationId) {
        notificationService.markAsRead(notificationId);
        return ResponseEntity.noContent().build();
    }

    @GetMapping("/countNotificationNoRead/{userId}")
    public ResponseEntity<Long> countNotificationNoRead(@PathVariable Long userId) {
        try {
            return ResponseEntity.ok(notificationService.countNotificationNoRead(userId));
        } catch (Exception e) {
            return ResponseEntity.internalServerError().build();
        }

    }

    // Método adicional para marcar todas las notificaciones como leídas
    @PutMapping("/mark-all-as-read/{userId}")
    public ResponseEntity<Void> markAllNotificationsAsRead(@PathVariable Long userId) {
        notificationService.markAllAsRead(userId);
        return ResponseEntity.noContent().build();
    }
}